keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,903 @@
+)
+```
+
+This approach has the advantage that the size of the `CSIStoragePool`
+objects do not depend on the number of nodes in the cluster for local
+storage.
+
+The downsides are:
+- Some attributes (driver name, topology) must be stored multiple times
+  compared to a single, more complex object that contains all information
+  for one CSI driver, so overall size in etcd is higher.
+- Higher number of objects which all need to be retrieved by a client
+  which does not already know which `CSIStoragePool` has a certain spec.
+- Constraints like "unique spec" cannot be enforced by the API server.
+
+##### Example: local storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-ab96d356-0d31-11ea-ade1-8b7e883d1af1
+spec:
+  driverName: hostpath.csi.k8s.io
+  nodes:
+  - node-1
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: <fallback>
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-c3723f32-0d32-11ea-a14f-fbaf155dff50
+spec:
+  driverName: hostpath.csi.k8s.io
+  nodes:
+  - node-2
+status:
+  classes:
+  - capacity: 512G
+    storageClassName: <fallback>
+```
+
+##### Example: affect of storage classes
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-9c17f6fc-6ada-488f-9d44-c5d63ecdf7a9
+spec:
+  driverName: lvm
+  nodes:
+  - node-1
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: striped
+  - capacity: 128G
+    storageClassName: mirrored
+```
+
+##### Example: network attached storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-b0963bb5-37cf-415d-9fb1-667499172320
+spec:
+  driverName: pd.csi.storage.gke.io
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-east-1
+status:
+  classes:
+  - capacity: 128G
+    storageClassName: <fallback>
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-64103396-0d32-11ea-945c-e3ede5f0f3ae
+spec:
+  driverName: pd.csi.storage.gke.io
+status:
+  classes:
+  - capacity: 128G
+    storageClassName: <fallback>
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-west-1
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: <fallback>
+```
+
+#### CSIDriver.spec.storageCapacity
+
+A new field `storageCapacity` of type `boolean` with default `false`
+in
+[CSIDriver.spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#csidriverspec-v1beta1-storage-k8s-io)
+indicates whether a driver deployment will create `CSIStoragePool`
+objects. This allows the scheduler to determine whether it needs to
+consider available capacity for volumes of that driver or can proceed
+without it (as it does now).
```

Comment:
How does it work for the last two use cases you listed earlier?  In those two use cases, driver wants to have CSIStoragePool but also wants to use a custom scheduler.

Reply:
In both cases the driver could leave the `CSIDriver.spec.storageCapacity` field unset, create `CSIStoragePool` objects, then let the custom scheduler work with those.

Reply:
I realized that the definition of `CSIDriver.spec.storageCapacity` was not quite clear on that. I enhanced it in 0fd7bc84.